### DISCUSSION TITLE
Error: file is encrypted or is not a database

### DISCUSSION THREAD

**Discussion by franck-serot** [sqlalchemy/sqlalchemy#6822] [Usage Questions]
### Describe the bug

Hi,
I created a simple Python app with 
- Flask==2.0.1
- Flask-SQLAlchemy==2.5.1
- SQLAlchemy==1.4.22
- sqlcipher3-binary==0.4.5
and the configuration 
- BASEDIR = os.path.abspath(os.path.dirname(__file__))
- SQLALCHEMY_DATABASE_URI = 'sqlite+pysqlcipher://:passphrase@/' + BASEDIR + '/db1.sqlite'

I tries to use sqlcipher 3.15.2 in order to check the content of the database with or without a passphrase.

Without the passphrase, the database is not ciphered and I can show the schema of the database for example.
With the passphrase, I defined the passphrase with the pragma command "PRAGMA key='passphrase';" but sqlcipher returns the following error :
   Error: file is encrypted or is not a database

I tried cipher a base directly with sqlcipher and I'm able to check the passphrase without no problem.

Would you please helpl me to understand the problem and solve it ?
Best regards,
Franck

### To Reproduce

```python
Without the passphrase, the database is not ciphered and I can show the schema of the database for example.
With the passphrase, I defined the passphrase with the pragma command "PRAGMA key='passphrase';" but sqlcipher returns the following error :
   Error: file is encrypted or is not a database
```

### Error

With sqlcipher : Error: file is encrypted or is not a database

### Versions

- OS: Ubuntu-20.04
- Python: 3.8
- SQLAlchemy: 1.4.22
- Database: 
- DBAPI: 

- Flask==2.0.1
- Flask-SQLAlchemy==2.5.1
- SQLAlchemy==1.4.22
- sqlcipher3-binary==0.4.5

**Comment by zzzeek** [comment]
what matters most is how this file got created.   

the program below will create an encrypted database file.   if i try to query it with raw sqlite3, I get "file is not a database", because it's encrypted.  But I can run the program repeatedly and it gets more rows back each time, by reading the file:

```py
from sqlalchemy import create_engine

e = create_engine(
    'sqlite+pysqlcipher://:some_passphrase@//var/tmp/db1.sqlite'
)

with e.connect() as conn:
    conn.exec_driver_sql("create table if not exists foo(id integer)")
    print("Existing rows: %s" % (conn.exec_driver_sql("select * from foo").all(), ))
    conn.exec_driver_sql("insert into foo (id) values (1)")
    print(conn.scalar("SELECT 1"))
```

**Comment by franck-serot** [comment]
I used SQLALCHEMY to create the database and encrypt it with sqlcipher3-binary in a Flask application.
Unless I don't understand, I should be able to use a tool like sqlcipher in a shell in order to connect to the database with the passphrase.
Why it's not the possible ?

**Comment by zzzeek** [comment]
> Why it's not the possible ?

no idea, SQLAlchemy does not implement this protocol, the sqlcipher library does so you'd need to explore it on that end why this happens.   

I can reproduce it not working, by running this script to create a new file

```py
import sqlcipher3 as sqlcipher

c = sqlcipher.connect('/var/tmp/db2.sqlite')

cursor = c.cursor()
cursor.execute("pragma key='some_passphrase'")

cursor.close()


cursor = c.cursor()
cursor.execute("create table if not exists foo(id integer)")
cursor.execute("insert into foo (id) values (1)")
cursor.close()

c.commit()

c.close()
```


then using sqlcipher:

```
$ sqlcipher  /var/tmp/db2.sqlite 
SQLCipher version 3.15.2 2016-11-28 19:13:37
Enter ".help" for instructions
Enter SQL statements terminated with a ";"
sqlite> pragma key='some_passphrase';
sqlite> select * from foo;
Error: file is encrypted or is not a database
```

you'd need to report a bug with pysqlcipher.

**Comment by CaselIT** [comment]
I think this could be closed or moved to a discussion, since I don't think there is nothing for sqlalchemy to do here
